File: objC/Forwarding.m
#import <Foundation/Foundation.h>

@interface B: NSObject {}
- (void) messageForB;
@end

@implementation B
- (void) messageForB {printf("Something \n");}
@end


@interface A : NSObject {
	B *b;
}
@end

@implementation A

- (id) init {
	if (self = [super init]) {
		b = [[B alloc] init];
	}
	return self;
}

- (void)forwardInvocation:(NSInvocation *) anInvocation {
    if ([b respondsToSelector: [anInvocation selector]])
        [anInvocation invokeWithTarget:b];
    else [super forwardInvocation:anInvocation];
}

- (NSMethodSignature *)methodSignatureForSelector:(SEL)aSelector {
    return [B instanceMethodSignatureForSelector:aSelector];
}

@end


int main(int argc, char const *argv[]) {
	A *a = [[A alloc] init];
	[a messageForB];
	return 0;
}

File: C++/Forwarding.cpp
#include <iostream>

using namespace std;

template <typename T>
class A {
public:
	A(T obj): _obj(obj) {}

	void messageForT() {_obj.messageForT();}

private:
	T _obj;
};

class B {
public:
	void messageForT() { cout << "Something\n";}
};

int main(int argc, char const *argv[])
{
	B b;
	A<B> a = A<B>(b);
	a.messageForT();
	return 0;
}